spec/controllers/omniauth_callbacks_controller_spec.rb

```diff
@@ -1,62 +1,136 @@
+    @request.env["devise.mapping"] = Devise.mappings[:user]
+    create(:org, managed: false, is_other: true)
+    @org = create(:org, managed: true)
+    @identifier_scheme = create(:identifier_scheme,
+                                name: 'openid_connect',
+                                description: 'CILogon',
+                                active: true,
+                                identifier_prefix: 'https://www.cilogon.org/')
+
+    # Mock OmniAuth data for OpenID Connect with necessary info
+    OmniAuth.config.mock_auth[:openid_connect] = OmniAuth::AuthHash.new({
+      provider: 'openid_connect',
+      uid: '12345',
+      info: {
+        email: 'test@example.com',
+        first_name: 'Test',
+        last_name: 'User',
+        name: 'Test User'
+      }
+    })
+
+    # Assign the mocked authentication hash to the request environment
+    @request.env["omniauth.auth"] = OmniAuth.config.mock_auth[:openid_connect]
+  end
+
+  describe 'GET #openid_connect' do
+    let(:auth) { request.env['omniauth.auth'] }
+    let!(:identifier_scheme) { IdentifierScheme.create(name: auth.provider) }
+
+    context 'when the email is missing and the user does not exist' do
+      before do
+        # Simulate missing email
+        OmniAuth.config.mock_auth[:openid_connect].info.email = nil
+        @request.env["omniauth.auth"] = OmniAuth.config.mock_auth[:openid_connect]
+      end
+
+      it 'redirects to the registration page with a flash message' do
+        get :openid_connect
+
+        expect(response).to redirect_to(new_user_registration_path)
+        expect(flash[:notice]).to eq('Something went wrong, Please try signing-up here.')
+      end
+    end
+
+    context 'when the user is not signed in but already exists' do
+      # let!(:user) { User.create(email: auth.info.email, password: 'password123') }
+      let!(:user) { User.create(email: 'test@example.com', firstname: 'Test', surname: 'User',  org: @org) }
+
+      before do
+        def User.from_omniauth(_auth)
+          User.find_by(email: 'test@example.com')
+        end
+      end
+
+      it 'signs in the existing user' do
+        get :openid_connect
+        # expect(subject.current_user).to eq(user)
+        expect(response).to redirect_to(root_path)
+        expect(flash[:notice]).to be_nil
+      end
+    end
+
+    context 'when the user is signed in and needs to link their OpenID Connect account' do
+      let!(:user) {  User.create(email: 'test@example.com', firstname: 'Test', surname: 'User',  org: @org) }
+      let(:current_user) { create(:user) }
+
+      before do
+        sign_in current_user
+
+        # Ensure from_omniauth returns nil, indicating no user is associated with the auth
+        # User.define_singleton_method(:from_omniauth) do |_auth|
+        #   nil
+        # end
+      end
+
+      it "links identifier to current user, sets flash notice, and redirects to root path" do
+        expect {
+          get :openid_connect
```

Comment:
I'm confused why `get :openid_connect` is working.
`get :openid_connect` seems to behave like `post :openid_connect` within these tests, but the same is not true outside of the tests.
![Screenshot from 2024-09-04 16-13-46](https://github.com/user-attachments/assets/94128505-f147-42c8-b203-846a968f6b53)
![Screenshot from 2024-09-04 16-13-11](https://github.com/user-attachments/assets/5783d3d5-326d-4913-8b55-50ef71e5915d)

Reply:
updated this